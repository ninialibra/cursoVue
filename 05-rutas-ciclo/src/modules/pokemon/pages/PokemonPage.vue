<template>
    <div>
        <h1>Pokemon: <span>#{{id}}</span></h1>
        <div v-if="pokemon">
            <img :src="pokemon.sprites.front_default" :alt="pokemon.name">
        </div>
    </div>
</template>

<script>
export default {
    props:{
        id:{
            type: Number,
            required: true
        }
    },
    data(){
        return{
            //id: this.$route.params.id
            pokemon: null
        }
    },
    created(){
        //console.log(this.$route);
        //const {id} = this.$route.params;
        //this.id=id;

        this.getPokemon();
    },
    methods:{
        async getPokemon(){

            try {
                
                const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`).then(r=>r.json());
                this.pokemon = pokemon;
                console.log(pokemon);

            } catch (error) {
                this.$router.push('/');
            }

        }
    },
    watch: {
        id(){
            this.getPokemon();
        }
    }
}
</script>