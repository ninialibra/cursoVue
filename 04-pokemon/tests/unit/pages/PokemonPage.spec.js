import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage';
import { mock_pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage Componen', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    });

    test('debe hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('debe de llamar mixPokemonArr al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        shallowMount(PokemonPage);

        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('debe de hacer match con el snapshot cuando cargan pokemons', () => {

        //el mount monta el snapshot al 100% con todo el ciclo de vida de los componentes
        const wrapper = mount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mock_pokemons,
                    pokemon: mock_pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                };
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('debe de mostrar los componentes de picture y options', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mock_pokemons,
                    pokemon: mock_pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                };
            }
        });

        const picture = wrapper.find('pokemon-picture-stub');
        const options = wrapper.find('pokemon-options-stub');

        expect(picture.exists()).toBeTruthy();
        expect(options.exists()).toBeTruthy();

        expect(picture.attributes('pokemonid')).toBe('1');
        expect(options.attributes('pokemons')).toBeTruthy();
    });

    test('pruebas con checkanswer', async() => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mock_pokemons,
                    pokemon: mock_pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                };
            }
        });

        await wrapper.vm.checkAnswer(1);

        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBe(true);
        expect(wrapper.find('h2').text()).toBe(`Correcto, ${mock_pokemons[0].name}!`);

        await wrapper.vm.checkAnswer(10);
        expect(wrapper.vm.message).toBe(`Ups! Era ${mock_pokemons[0].name}`);

    });
});