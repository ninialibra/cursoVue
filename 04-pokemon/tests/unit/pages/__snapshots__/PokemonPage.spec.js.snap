// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Componen debe de hacer match con el snapshot cuando cargan pokemons 1`] = `
<div>
  <div>
    <h1>¿Quién es este Pokémon?</h1>
    <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="Pokemon" class="hidden-pokemon"></div>
    <div class="options-container">
      <ul>
        <li>bulbasaur</li>
        <li>ivysaur</li>
        <li>venusaur</li>
        <li>charmander</li>
      </ul>
    </div>
    <!--v-if-->
  </div>
</div>
`;

exports[`PokemonPage Componen debe hacer match con el snapshot 1`] = `
<div>
  <h1>Espere por favor...</h1>
</div>
`;
